### PR TITLE
Fix C# compiler generated Linq selectors to use anonymous object crea…

### LINKED ISSUE #25268: Compiler generated selectors for Linq queries in C# use regular IObjectCreationOperation instead of IAnonymousObjectCreationOperation
Examine shape of SelectMany selector in base-line for existing unit-test TransparentIdentifiers_FromLet_IOperation in src\Compilers\CSharp\Test\Semantic\Semantics\QueryTests.cs

```
 IObjectCreationOperation (Constructor: <anonymous type: System.Int32 x, System.Int32 y>..ctor(System.Int32 x, System.Int32 y)) (OperationKind.ObjectCreation, Type: <anonymous type: System.Int32 x, System.Int32 y>, IsImplicit) (Syntax: 'from int y in c2')
    Arguments(2):
        IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: x) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: 'from int y in c2')
          IParameterReferenceOperation: x (OperationKind.ParameterReference, Type: System.Int32, IsImplicit) (Syntax: 'from int y in c2')
          InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
          OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
        IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: y) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: 'from int y in c2')
          IParameterReferenceOperation: y (OperationKind.ParameterReference, Type: System.Int32, IsImplicit) (Syntax: 'from int y in c2')
          InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
          OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
    Initializer: 
    null
 ```

Observed:
Even though this represents creation of an instance of an Anonymous Type, the node is  IObjectCreationOperation rather than an IAnonymousObjectCreationOperation. Semantically this is accurate, but some useful information is not exposed and left for the consumer to figure out. In particular, it is hard to say what properties are being initializer with values passed as constructor arguments. Also, a simple direct serialization of the IObjectCreationOperation into syntax will never produce valid syntax because Anonymous Types cannot be explicitly named and, therefore, their constructors cannot be explicitly invoked. The only valid way to represent this in syntax is to use an Anonymous Type creation expression, which is usually mapped to IAnonymousObjectCreationOperation.

Expected: 
IAnonymousObjectCreationOperation node. BTW, this is what happens for VB.

I believe this issue is blocking for "Linq query to method calls" refactoring feature.

CC @dotnet/analyzer-ioperation, @ivanbasov

### PR DESCRIPTION
**Pull request by mavasani**
…tion instead of regular object creation

Fixes #25268

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Query.cs [hunk 1]
```diff
@@ -785,11 +785,10 @@ protected BoundCall MakeQueryInvocation(CSharpSyntaxNode node, BoundExpression r
 
         protected BoundExpression MakeConstruction(CSharpSyntaxNode node, NamedTypeSymbol toCreate, ImmutableArray<BoundExpression> args, DiagnosticBag diagnostics)
         {
-            AnalyzedArguments analyzedArguments = AnalyzedArguments.GetInstance();
-            analyzedArguments.Arguments.AddRange(args);
-            var result = BindClassCreationExpression(node, toCreate.Name, node, toCreate, analyzedArguments, diagnostics);
+            Debug.Assert(toCreate.IsAnonymousType);
+
+            var result = new BoundAnonymousObjectCreationExpression(node, toCreate.InstanceConstructors[0], args, ImmutableArray<BoundAnonymousPropertyDeclaration>.Empty, toCreate);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Query.cs]: This is potentially cutting out a good deal of error checking done in BindClassCreationExpression. Are we sure this is the right change to make?

Comment (mavasani) [reply src/Compilers/CSharp/Portable/Binder/Binder_Query.cs]: I was hoping unit tests to catch any missing functionality, if it was indeed desired or possible to hit those cases. No C# compiler unit tests failed locally for me after this change.

Comment (mavasani) [reply src/Compilers/CSharp/Portable/Binder/Binder_Query.cs]: Adding @gafter @jcouv to confirm.

**Review by 333fred — CHANGES_REQUESTED** [review]
Given how late we are in the cycle for 15.8, I'm very uncomfortable with this change, as it can definitely have far-reaching impacts on how LINQ queries are generated. I think we need to come up with a different way of handling this.

**Comment by 333fred** [comment]
I don't have any specific test cases off the top of my head. However, the LocalRewriter has certain assumptions about input and will not perform the error checking that BindClassCreationExpression does. My point is that, for changing LINQ binding, we need to have a more concrete belief that this is the right change than the assertion that none of our tests broke, especially at this point in the cycle for 15.8.

**Comment by AlekseyTs** [comment]
@mavasani 
I think we can reduce the risk if we move the change from the binder to the IOperation factory. A user cannot  new an anonymous type with a name. So, if we see a regular object creation that creates an anonymous type, it should always be safe to surface it as an anonymous object creation.

**Comment by mavasani** [comment]
@AlekseyTs Thanks. I will file a follow-up 16.0 issue for the compiler team to remove the operation factory workaround and attempt to fix the binder or add relevant unit tests.

File: src/Compilers/CSharp/Portable/Binder/Binder_Query.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Binder/Binder_Query.cs]: >  No C# compiler unit tests failed locally for me after this change.

That is not good evidence of a change being correct when the change itself is a bit speculative. It's more likely to indicate there is a test gap.

Comment (mavasani) [reply src/Compilers/CSharp/Portable/Binder/Binder_Query.cs]: Filed https://github.com/dotnet/roslyn/issues/28157 for the compiler team to make the correct change and unit tests at a later milestone. I will update this PR to put a workaround in operation factory as suggested by @AlekseyTs

**Comment by mavasani** [comment]
@dotnet-bot retest ubuntu_16_mono_debug_prtest please

**Review by 333fred — APPROVED** [review]
Thanks Manish. Much more comfortable with this implementation.

**Review by jcouv — APPROVED** [review]
LGTM Thanks

**Comment by jcouv** [comment]
@jaredpar for ask-mode approval for 15.8. Thanks

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest ubuntu_16_mono_debug_prtest please

**Comment by mavasani** [comment]
Ping @jaredpar @jinujoseph for ask mode approval

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview5

**Pull request merged**